### ISSUE TITLE
doubt num of num_threads is different to  linux (ps -ef |grep python train.py|wc -l)

### ISSUE THREAD

**Issue by daixiangzi** [NVIDIA/DALI#5137]
### Describe the question.

in fact，my pc:cpu_num =128
I only use torchrun and ddp ,6 gpu on one machine, however,i get 194 bypass run ps -ef|grep python train.py |wc -l
I think it should thread = num_gpu*workers=24,
`
pipe = Pipeline(
        batch_size=batch_size,
        num_threads=4,
        device_id=local_rank % 6,
        prefetch_queue_depth=3,
        seed=rank + 1467,
    )
`

### Check for duplicates

- [X] I have searched the [open bugs/issues](https://github.com/NVIDIA/DALI/issues) and have found no duplicates for this bug report

**Comment by daixiangzi** [comment]
others,i set num_threads=10,but I get 194 by ps -ef|grep python train.py |wc -l

**Comment by JanuszL** [comment]
Hi @daixiangzi,

Thank you for reaching out.
DALI uses the number of threads value in many different ways, depending on the operators you may have in your pipeline.
Cases that come to my mind where this value is used (but this is not limited only to items listed below):
- it creates N threaded pool for the CPU work
- mixed image decoder will create N threaded pool of workers just for the image decoding
- selected readers - like the numpy reader may use an separated thread pool to speed up the data reading
- the DALI executor creates one thread per stage of work - CPU, mixed, and GPU (if you ask to select asynchronous execution - that is the default option during the creation of the pipeline)

So long story short, it may happen that DALI will use in total more thread than the value provided to `num_threads`.
Do you see any particular performance issue coming from that fact?

**Comment by daixiangzi** [comment]
this my Dali code about pipeline:
`device_memory_padding = 211025920
    host_memory_padding = 140544512
    preallocate_width_hint = 5980
    preallocate_height_hint = 6430 
 with pipe:
        jpegs, labels = fn.readers.mxnet(
            path=rec_file,
            index_path=idx_file,
            initial_fill=16384,
            num_shards=world_size,
            shard_id=rank,
            random_shuffle=True,
            pad_last_batch=False,
            name="train",
            stick_to_shard=True
        )
        if is_training:
            images = fn.decoders.image_random_crop(
                jpegs,
                device="mixed",
                output_type=types.RGB,
                device_memory_padding=device_memory_padding,
                host_memory_padding=host_memory_padding,
                random_aspect_ratio=[0.8, 1.25],
                random_area=[0.7, 1.0],
                num_attempts=100,
                hw_decoder_load=0.75,
                preallocate_width_hint=preallocate_width_hint,
                preallocate_height_hint=preallocate_height_hint,
            )
            images = fn.resize(
                images,
                device="gpu",
                resize_x=image_size[0],
                resize_y=image_size[1],
                interp_type=types.INTERP_LINEAR,
            )
            mirror = fn.random.coin_flip(probability=0.5)
        else:
            images = fn.decoders.image(jpegs, device="mixed", output_type=types.RGB)
            images = fn.resize(
                images,
                device="gpu",
                size=int(256 / 224 * image_size[0]),
                mode="not_smaller",
                interp_type=types.INTERP_LINEAR,
            )
            mirror = False

        images = fn.crop_mirror_normalize(
            images.gpu(),
            dtype=types.FLOAT,
            output_layout="CHW",
            crop=(image_size[0], image_size[1]),
            mean=mean,
            std=std,
            mirror=mirror)
        pipe.set_outputs(images, labels)
    pipe.build()`

**Comment by daixiangzi** [comment]
> Hi @daixiangzi,
> 
> Thank you for reaching out. DALI uses the number of threads value in many different ways, depending on the operators you may have in your pipeline. Cases that come to my mind where this value is used (but this is not limited only to items listed below):
> 
> * it creates N threaded pool for the CPU work
> * mixed image decoder will create N threaded pool of workers just for the image decoding
> * selected readers - like the numpy reader may use an separated thread pool to speed up the data reading
> * the DALI executor creates one thread per stage of work - CPU, mixed, and GPU (if you ask to select asynchronous execution - that is the default option during the creation of the pipeline)
> 
> So long story short, it may happen that DALI will use in total more thread than the value provided to `num_threads`. Do you see any particular performance issue coming from that fact?

this my codes：I use mxnet rec file
```python
device_memory_padding = 211025920
host_memory_padding = 140544512
preallocate_width_hint = 5980
preallocate_height_hint = 6430

with pipe:
    jpegs, labels = fn.readers.mxnet(
        path=rec_file,
        index_path=idx_file,
        initial_fill=16384,
        num_shards=world_size,
        shard_id=rank,
        random_shuffle=True,
        pad_last_batch=False,
        name="train",
        stick_to_shard=True
    )
    if is_training:
        images = fn.decoders.image_random_crop(
            jpegs,
            device="mixed",
            output_type=types.RGB,
            device_memory_padding=device_memory_padding,
            host_memory_padding=host_memory_padding,
            random_aspect_ratio=[0.8, 1.25],
            random_area=[0.7, 1.0],
            num_attempts=100,
            hw_decoder_load=0.75,
            preallocate_width_hint=preallocate_width_hint,
            preallocate_height_hint=preallocate_height_hint,
        )
        images = fn.resize(
            images,
            device="gpu",
            resize_x=image_size[0],
            resize_y=image_size[1],
            interp_type=types.INTERP_LINEAR,
        )
        mirror = fn.random.coin_flip(probability=0.5)
    else:
        images = fn.decoders.image(jpegs, device="mixed", output_type=types.RGB)
        images = fn.resize(
            images,
            device="gpu",
            size=int(256 / 224 * image_size[0]),
            mode="not_smaller",
            interp_type=types.INTERP_LINEAR,
        )
        mirror = False

    images = fn.crop_mirror_normalize(
        images.gpu(),
        dtype=types.FLOAT,
        output_layout="CHW",
        crop=(image_size[0], image_size[1]),
        mean=mean,
        std=std,
        mirror=mirror
    )
    pipe.set_outputs(images, labels)

pipe.build()
```

**Comment by daixiangzi** [comment]
As the training progresses, the speed becomes slower and slower

**Comment by JanuszL** [comment]
@daixiangzi,

I'm afraid that the `the speed becomes slower and slower` is caused by the disc cache, not the DALI itself.
You can read about this more [here](https://github.com/NVIDIA/DALI/issues/3473) or [here](https://github.com/NVIDIA/DALI/issues/3348).

**Comment by daixiangzi** [comment]
> @daixiangzi,
> 
> I'm afraid that the `the speed becomes slower and slower` is caused by the disc cache, not the DALI itself. You can read about this more [here](https://github.com/NVIDIA/DALI/issues/3473) or [here](https://github.com/NVIDIA/DALI/issues/3348).
thank you ,I understand,but above train_thread greater than dali_thread_num how to explain

**Comment by daixiangzi** [comment]
> @daixiangzi,
> 
> I'm afraid that the `the speed becomes slower and slower` is caused by the disc cache, not the DALI itself. You can read about this more [here](https://github.com/NVIDIA/DALI/issues/3473) or [here](https://github.com/NVIDIA/DALI/issues/3348).

thank you ,I understand,but above train_thread greater than dali_thread_num how to explain

**Comment by JanuszL** [comment]
> thank you ,I understand,but above train_thread greater than dali_thread_num how to explain

Can you capture a run profile using [NSight system](https://developer.nvidia.com/nsight-systems) so we can see all the threads and check which one belongs to DALI (they should be annotated in the profile)?
